### ISSUE TITLE
bar3 issues with manual solution and examples

### ISSUE THREAD

**Issue by elektronxyz** [matlab2tikz/matlab2tikz#878]
I wanted to produce some bar3 plots with colorbar. To avoid unprintable documents a decided to use the shader=flat corner option.

Here is an matlab example:

`
[fig] = figure('Name','Example bar3','Position',[50 50 800 523]);

for x = [1:1:3]
    for y = [1:1:3]
        data(x,y) = 10 - x * y;
    end
end

box on;
grid on;

hbar = bar3(data);
view([52.5, 30]);

colormap('jet');
hcol = colorbar;

for k = 1:length(hbar)
    zdata = hbar(k).ZData;
    hbar(k).CData = zdata;
    hbar(k).FaceColor = 'flat';
end
`

Producing the following figure:
![example](https://cloud.githubusercontent.com/assets/15018040/14282795/55acbda4-fb40-11e5-8b59-94d4b80861bb.png)

The current development version of Matlab2tikz produces the following code:
[example.tikz.txt](https://github.com/matlab2tikz/matlab2tikz/files/204511/example.tikz.txt)

The compiled picture is not usable.
![example-tikz](https://cloud.githubusercontent.com/assets/15018040/14282999/80a62c4c-fb41-11e5-9ab8-f9d677054c4a.png)

Therefore, I removed the z butter statements with the following result.
![example-tikz-without-z-buffer](https://cloud.githubusercontent.com/assets/15018040/14283029/a5080c68-fb41-11e5-8b21-f07309523128.png)

To get the desired results, I had to place some entries with `nan`. 
Please have a look in the tikz file.
[example-edit-working.tikz.txt](https://github.com/matlab2tikz/matlab2tikz/files/204516/example-edit-working.tikz.txt)
  
Now, it looks like the matlab example.
![example-tikz-without-z-buffer-and-removed-entries](https://cloud.githubusercontent.com/assets/15018040/14283087/f044d990-fb41-11e5-97d6-61ec8b577477.png)

This manual solution is not adequate for large pictures. Therefore, I would like to ask, if it is possible to change matlab2tikz in the proposed manner.

Best regards,
  Elektron

**Comment by miscco** [comment]
I started improving bar code in #851. I will have a look at this when the little one sleeps

**Comment by egeerardyn** [comment]
Thank you for this elaborate bug report!

If I interpret your figures correctly, you have removed the ground planes of all those boxes. That is something that should be feasible on our end, I think.
The other thing causing problems is the back plane of those boxes, identifying those could be bit harder, but should be possible as well.

@miscco: The 2D and 3D `bar` code is quite different if I'm not mistaken. For 3D, I think (at least for HG1 and Octave), we convert the raw underlying patch objects into TikZ code. For HG2, I suspect we will have to duplicate quite a bit of code to filter out these unwanted surfaces.

**Comment by miscco** [comment]
I have the very strong feeling, that this is actually an ordering problem. From the pgfplots manual section 4.6.5 on page 135 I find the following options for z buffer=:

```
/pgfplots/z buffer=default | none | auto | sort | reverse x seq | reverse y seq |reverse xy seq
```

Given that in the example the y-Axis is reversed i would guess that we would have to use z buffer = reverse y seq.

I will test that once i have access to matlab

**Comment by elektronxyz** [comment]
I tried your suggestion to use 'buffer=reverse y seq'.

Here is the result:
![example-tikz-reverse-y-seq](https://cloud.githubusercontent.com/assets/15018040/14284955/01f7b326-fb4a-11e5-8a40-9e6cb13c4545.png)

 'buffer=reverse x seq' is not better.

Best regards,
  Elektron

**Comment by elektronxyz** [comment]
I also tried the other z buffer options. none and auto produce the same result as without z buffer.

reverse xy seq is also not usable.

**Comment by miscco** [comment]
Yeah, i dont know why, but it just seems, that the ordering of the individual patches is off, even with reverse odering. I think it might be best to switch to z buffer=auto and then remove the ground plane?

**Comment by elektronxyz** [comment]
@egeerardyn: As far as I understood what I did, I removed the ground planes. This should help for all rotations, but should only be applied when the view angle is positive. 

If you can give me a hint, where in matlab2tikz the bar3 function is handled, I can try to fix is by myself.

**Comment by elektronxyz** [comment]
I found the following commend in the code:

>     % - hist3D plots should not be z-sorted or the highest bars will cover
    %   the shortest one even if positioned in the  back

As the hist3 command seems to work properly, it might be an option to use the same code for bar3.

Best regards
  Elektron

**Comment by egeerardyn** [comment]
@elektronxyz That seems like a good idea.

Probably you can get that part working already by replacing the line
```matlab
isHist3D     = strcmpi(get(h,'tag'), 'hist3');
```
with
```matlab
isHist3D     = ismember(get(h,'tag'), {'hist3', 'bar3'});
```

Removing the ground planes still needs to be done, though.

**Comment by elektronxyz** [comment]
Thanks a lot. The fix for the z buffer works great. 

Any advice, where to look for the ground planes ?

**Comment by egeerardyn** [comment]
For the ground planes, I think it's going to be a bit more complicated. I suppose that in `drawSurface`, we will have to check for the `Tag` being `bar3` or something and then remove the ground planes (probably `z==0`) from the `{XYZC}Data` of the plot. It's not going to be a one-line fix.

**Comment by miscco** [comment]
I think one could add

```matlab
    % Remove the ground plane in a 3d bar plot
    % NOTE: Issue #878
    if(strcmpi(get(h, 'Tag'), 'bar3'))
       id_zeroPlane = data(:, 3) == 0;
       data(id_zeroPlane,:) = [];
    end
```

After the assignment of 

```matlab
    data = [data, color];
```

in drawSurface

**Comment by egeerardyn** [comment]
@miscco Looks about right. I, however, don't see the point in setting them to NaN instead of removing: it just fills the LaTeX buffers but doesn't add anything...

**Comment by miscco** [comment]
@egeerardyn: Actually at second thought it doesnt really solve the problem by any stretch. 

The first mistake is that it assumes the ground plane to be at 0, which is not always true, i think there is a basevalue. One could avoid that by switching to min/max which leads to:

The second mistake is that it does not account for the direction of the axis. If for example zDir is reversed we have to delete the maximal z value and vice versa. However, this leads to 

The third mistake, it does not account for the view. Depending on the orientation different surfaces are visible. This bundles with the direction of the axes, but is its own problem.

**Comment by egeerardyn** [comment]
@miscco I agree with your concerns, viewing angle and so on has to be accounted for. When I see that, perhaps it is more prudent to carry out those simplifications in `cleanfigure`. 

Eventually, it might even pay off to do some kind of ray-tracing in `cleanfigure` such that we can detect invisible elements and remove the from the figure (the general approach is going to be a lot of work, though, as that would basically mean that we have to quite some 3D rendering work).


Just as a remainder for myself (and hopefully it doesn't bite anyone):
Something that struck me as odd in this MWE, is that the back planes are all in that dark blue (just rotate the figure to see). So I suppose the `hbar` handles are only the front planes. Standard `bar3` plots have those planes in the normal colors, though.

**Comment by miscco** [comment]
Actually i think everything is in place for ray tracking. As far as i understand this is handled by the 3D transformation matrix, that we utilize in getVisualData. 

The odd color also struck me

**Comment by egeerardyn** [comment]
@miscco It's a bit more than that. Those transformation matrices don't tell whether an object is actually visible or not. They indicate where such a 3D object would be mapped to on the 2D viewscape if it were the only object in existence. Here, the problem is that LaTeX has a hard time figuring out what object should be in front and hence what all those things together should look like.

**Comment by miscco** [comment]
@egeerardyn i thought the 4th component is the distance from the camera. But i have to check that further

**Comment by egeerardyn** [comment]
@miscco I don't know a lot about 4-vector calculations, so I would have to read up on those as well (I do know, however, that first 3 are typical 3D coordinates and indeed the 4th is to allow translations). So it might very well be the distance to the camera.

**Comment by elektronxyz** [comment]
@miscco: Thanks for the hint to have a look into the drawSurface function on line 2960.

Just removing the lines does not work. They have to be replaced by nan entries.

When I modify your patch to replace the ground point by nan entries, no cubes are visible.
According to my tests, only the ground points with the highest distance from the view point have to be replaced. In my case, the have to lowest x-, y- and z- coordinates. 

I implement a stupid pattern matching to overwrite these points by nan entries. Its ugly code, but it works in my case ;-) 

`    % Remove one point of the ground plane in a 3d bar plot
    % NOTE: Issue #878
     if(strcmpi(get(h, 'Tag'), 'bar3'))
       backX = min(data(:, 1));
       backZ = min(data(:, 3));   
       for i = 1:(size(data, 1) - 4)
          if and(and(data(i, 1) == backX, data(i + 1, 3) > backZ), data(i + 2, 3) > backZ )
              data(i, :) = NaN;
              data(i + 4, :) = NaN;
          end
       end
     end`

Unfortunately, I have no idea how to read out the view angle ( view([52.5, 30]); ) in this function to implement a more general solution. Therefore, any help would be appreciate to add support for bar3 into matlab2tikz.

**Comment by okomarov** [comment]
@elektronxyz You can use `[az,el] = view(m2t.currentHandles.gca)`

**Comment by miscco** [comment]
@elektronxyz You can have a look at cleanfigure, and there getVisualData, wehere we calculate the 2D projection of 3D data.

**Comment by elektronxyz** [comment]
It is not so easy as I thought :-(

I tested examples for all quadrants of view angles. While the first quadrant works by replacing some point with nan's, the order of objects is wrong for the other quadrants.
 
To get a good picture of the third quadrant from 180°-270° the order of the child objects in line 645 has to be inverted. In this case, it is not necessary to replace points. But I'm not sure if it is a good idea to build in a special case for bar3 and this view angle at this level of the program.  

original: for child = children(end:-1:1)' --> does not work
new: for child = children(1:1:end)' --> works 

Best regards,
  Elektron

**Comment by elektronxyz** [comment]
Thanks to okomarov, the code is only applied for the first quadrant and positive view angels.

    % Remove one point of the ground plane in a 3d bar plot
    % NOTE: Issue #878
    if(strcmpi(get(h, 'Tag'), 'bar3'))
        [az,el] = view(m2t.currentHandles.gca);
        if and(and(az >= 0, az <= 90), el > 0)
            backX = min(data(:, 1));
            backZ = min(data(:, 3));   
            for i = 1:(size(data, 1) - 4)
                if and(and(data(i, 1) == backX, data(i + 1, 3) > backZ), data(i + 2, 3) > backZ )
                    data(i, :) = NaN;
                    data(i + 4, :) = NaN;
                end
            end
        end
    en

**Comment by klands** [comment]
I'm having the same issue as thoroughly described by elektronxyz, but without the acumen to manually fix it myself without serious effort. Has any progress been made on this topic that could help me?
Thanks!
klands
